src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableDefaultConverter.cs

```diff
@@ -41,6 +41,10 @@ protected static JsonConverter<TElement> GetElementConverter(ref WriteStack stat
             [MaybeNullWhen(false)] out TCollection value)
         {
             JsonTypeInfo elementTypeInfo = state.Current.JsonTypeInfo.ElementTypeInfo!;
+            if (!state.Current.JsonTypeInfo.NumberHandlingIsApplicableToElementType)
+            {
+                state.Current.NumberHandling = null;
```

Comment:
This approach doesn't seem right to me, current number handling is [explicitly inherited by child nodes](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs#L152) so erasing it here might mean we're serializing things incorrectly downstream.

I might be wrong, but there should be a test ensuring that this doesn't happen.

Reply:
cc @layomia who knows more about this.

Reply:
Yeah we shouldn't do this. @devsko can you consider bringing in the changes from https://github.com/dotnet/runtime/pull/42239 which also fixed the issue this PR solves? It is based on adding a check for `IsInternalConverterForNumberType` in the right places. You don't need to address feedback from https://github.com/dotnet/runtime/pull/42239#discussion_r494031684 which requires articulation beyond the scope of this PR.